### PR TITLE
jerry: add heap profiler

### PR DESCRIPTION
**Pull request by algebrait**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
  - test/run_pass/test_profilerjs
- [x] documentation is changed or added
  - docs/api/Profiler.md
  - docs/devs/Optimization-Tips.md
- [x] object node
- [x] string node
- [x] number node
- [x] bytecode node
- [x] synthetic node
- [x] edge in mark-sweep gc's reference

### CODE REVIEW

File: deps/jerry/jerry-core/api/jerry.c [hunk 1]
```diff
@@ -3629,6 +3630,26 @@ jerry_stop_cpu_profiling (void)
 #endif
 }
 
+void
+jerry_heap_profiler_take_snapshot (const char *path)
+{
+  jerry_assert_api_available ();
+
+#ifndef JERRY_HEAP_PROFILER
+  JERRY_UNUSED (path);
```
Comment (yorkie) [inline deps/jerry/jerry-core/api/jerry.c]: If `JERRY_HEAP_PROFILER` not defined, we should not do operate on any file handle.

File: src/modules/iotjs_module_profiler.c [hunk 2]
```diff
@@ -1,5 +1,14 @@
 #include "iotjs_def.h"
 
+JS_FUNCTION(TakeSnapshot) {
+  if (!jerry_value_is_string(jargv[0]))
+    return JS_CREATE_ERROR(COMMON, "filepath should be required.");
+
+  iotjs_string_t filepath = JS_GET_ARG(0, string);
+  jerry_heap_profiler_take_snapshot ((const char*)iotjs_string_data(&filepath));
+  return jerry_create_boolean(true);
```
Comment (yorkie) [inline src/modules/iotjs_module_profiler.c]: Call `iotjs_string_destroy(&filepath)` is required before return.

File: src/modules/iotjs_module_profiler.c [hunk 3]
```diff
@@ -1,5 +1,14 @@
 #include "iotjs_def.h"
 
+JS_FUNCTION(TakeSnapshot) {
+  if (!jerry_value_is_string(jargv[0]))
+    return JS_CREATE_ERROR(COMMON, "filepath should be required.");
+
+  iotjs_string_t filepath = JS_GET_ARG(0, string);
+  jerry_heap_profiler_take_snapshot ((const char*)iotjs_string_data(&filepath));
```
Comment (yorkie) [inline src/modules/iotjs_module_profiler.c]: Remove the space after the function `jerry_heap_profiler_take_snapshot`, that might be only in the JerryScript source.

File: deps/jerry/tools/snap.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
+var fs = require('fs');
```
Comment (yorkie) [inline deps/jerry/tools/snap.js]: Strict mode is required.

Comment (yorkie) [reply deps/jerry/tools/snap.js]: Where should this file be? The CPU profiling tool are https://github.com/Rokid/ShadowNode/tree/master/tools/profiler.

File: deps/jerry/tools/snap.js [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+const NODE_FILED_OFFSET_TYPE = 0;
+const NODE_FIELD_OFFSET_NAME = 1;
+const NODE_FIELD_OFFSET_ID = 2;
+const NODE_FIELD_OFFSET_SELF_SIZE = 3;
+
+const EDGE_FIELD_OFFSET_TYPE = 0;
+const EDGE_FIELD_OFFSET_NAME = 1;
+const EDGE_FIELD_OFFSET_TO_NODE = 2;
+
+const DUMMY_STRING = "<dummy>";
```
Comment (yorkie) [inline deps/jerry/tools/snap.js]: Use the single quote for strings always in JavaScript.

File: deps/jerry/tools/snap.js [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    console.log(JSON.stringify(this.edgesOutput));
+    console.log(",\n\"trace_function_infos\":[],\n\"trace_tree\":[],\n\"samples\":[],\n\"strings\":[");
+
+    console.log("\"<dummy>\"");
+    for(var i = 0; i < this.jerrySnapshot.strings.length; i++) {
+      var stringItem = this.jerrySnapshot.strings[i];
+      console.log(","+JSON.stringify(stringItem.chars));
+    }
+
+    console.log("]}");
```
Comment (yorkie) [inline deps/jerry/tools/snap.js]: I would think here a more readable generator based on JSON is needed :)

File: deps/jerry/jerry-core/api/jerry.c (hunk 1 above)
Comment (algebrait) [reply deps/jerry/jerry-core/api/jerry.c]: #define JERRY_UNUSED(x) ((void) (x))
JERRY_UNUSED does nothing on it's param. It just suppresses compiler warning "define but not used".

File: deps/jerry/tools/snap.js (hunk 4 above)
Comment (algebrait) [reply deps/jerry/tools/snap.js]: I would better place CPU and heap profiler into jerry source tree, because it is not related to shadow-node. The CPU profiler scripts will also be moved in other PR.

File: deps/jerry/tools/snap.js (hunk 5 above)
Comment (algebrait) [reply deps/jerry/tools/snap.js]: 3x!
The eslint has already give me many warning. I'll fix them all.

File: deps/jerry/tools/snap.js (hunk 6 above)
Comment (algebrait) [reply deps/jerry/tools/snap.js]: 3x! I will learn it.

**Comment by algebrait** [comment]
My understanding of strings in heap is wrong.
JERRY_CONTEXT (string_list_first_p) is only list of parser literal strings.
It seems there is no simple iterator of all strings.
The heap dumper and converter should be re-implemented.

File: deps/jerry/jerry-core/ecma/base/ecma-alloc.c [hunk 7]
```diff
@@ -28,8 +28,10 @@ JERRY_STATIC_ASSERT (((sizeof (ecma_property_value_t) - 1) & sizeof (ecma_proper
 JERRY_STATIC_ASSERT (sizeof (ecma_string_t) == sizeof (uint64_t),
                      size_of_ecma_string_t_must_be_less_than_or_equal_to_8_bytes);
 
+#ifndef JERRY_HEAP_PROFILER
 JERRY_STATIC_ASSERT (sizeof (ecma_extended_object_t) - sizeof (ecma_object_t) <= sizeof (uint64_t),
                      size_of_ecma_extended_object_part_must_be_less_than_or_equal_to_8_bytes);
+#endif /* JERRY_HEAP_PROFILER */
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/base/ecma-alloc.c]: Why do we disable this assertion in heap profiling?

File: deps/jerry/jerry-core/ecma/base/ecma-gc.c [hunk 8]
```diff
@@ -91,8 +91,23 @@ ecma_gc_is_object_visited (ecma_object_t *object_p) /**< object */
...
+                            ecma_string_t *edge_name_p, /**< edge name */
+                            v8_edge_type_t edge_type, /**< v8 edge type */
+                            void *args) /**< callback args */
 {
+  JERRY_UNUSED (from);
+  JERRY_UNUSED (edge_name_p);
+  JERRY_UNUSED (edge_type);
+  JERRY_UNUSED (args);
+
+  if (!ecma_is_value_object(to))
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/base/ecma-gc.c]: I guess this might have a coding style issue.

File: deps/jerry/jerry-core/ecma/base/ecma-gc.c [hunk 9]
```diff
@@ -165,11 +183,21 @@ ecma_gc_mark_property (ecma_property_pair_t *property_pair_p, /**< property pair
 
       ecma_value_t value = property_pair_p->values[index].value;
 
-      if (ecma_is_value_object (value))
+      // if (ecma_is_value_object (value))
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/base/ecma-gc.c]: If a statement is not used, how about removing from source or add a TODO/FIXME?

File: deps/jerry/jerry-core/ecma/base/ecma-gc.c [hunk 10]
```diff
@@ -165,11 +183,21 @@ ecma_gc_mark_property (ecma_property_pair_t *property_pair_p, /**< property pair
...
+        ecma_string_t *prop_name_p = ecma_string_from_property_name (property,
+            property_pair_p->names_cp[index]);
+        if (prop_name_p)
+        {
+          callback (object_p, value, prop_name_p, EDGE_TYPE_PROPERTY, callback_args);
+          ecma_deref_ecma_string (prop_name_p);
+        }
+        else
+        {
+          // TODO: when happen?
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/base/ecma-gc.c]: It's better that wrap the author name within TODO :)

```c
// TODO(algebrait): when happen?
```

File: deps/jerry/jerry-core/ecma/builtin-objects/ecma-builtins.inc.h [hunk 11]
```diff
@@ -296,7 +296,7 @@ BUILTIN_ROUTINE (ECMA_BUILTIN_ID_TYPE_ERROR_THROWER,
                  ECMA_BUILTIN_ID_FUNCTION_PROTOTYPE,
                  false,
                  type_error_thrower,
-                 LIT_MAGIC_STRING_TYPE_ERROR_UL)
+                 LIT_MAGIC_STRING_ERROR_UL)
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/builtin-objects/ecma-builtins.inc.h]: This should still be LIT_MAGIC_STRING_TYPE_ERROR_UL?

File: deps/jerry/jerry-core/ecma/base/ecma-helpers.h [hunk 12]
```diff
@@ -377,6 +377,47 @@ lit_utf8_size_t ecma_number_to_utf8_string (ecma_number_t num, lit_utf8_byte_t *
...
+typedef enum
+{
+  EDGE_TYPE_CONTEXT = 0,
+  EDGE_TYPE_ELEMENT = 1,
+  EDGE_TYPE_PROPERTY = 2,
+  EDGE_TYPE_INTERNAL = 3,
+  EDGE_TYPE_HIDDEN = 4,
+  EDGE_TYPE_SHORTCUT = 5,
+  EDGE_TYPE_WEAK = 6,
+} v8_edge_type_t;
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/base/ecma-helpers.h]: Strictly to say, names start with v8 in ecma is a bit weird.

File: deps/jerry/jerry-core/ecma/operations/ecma-objects.c [hunk 13]
```diff
@@ -1889,6 +1889,165 @@ ecma_object_class_is (ecma_object_t *object_p, /**< object */
...
+    {
+      break;
+    }
+    case ECMA_OBJECT_TYPE_BOUND_FUNCTION:
+    {
+      break;
+    }
+    case ECMA_OBJECT_TYPE_PSEUDO_ARRAY:
+    {
+      break;
```
Comment (yorkie) [inline deps/jerry/jerry-core/ecma/operations/ecma-objects.c]: Could we just return magic strings like what we have done on LIT_MAGIC_STRING_EXTERNAL_FUNCTION and CONFIG_DISABLE_ES2015_ARROW_FUNCTION? This might be useful data for heap profiler :)

File: deps/jerry/jerry-core/include/jerryscript-core.h [hunk 14]
```diff
@@ -537,6 +537,8 @@ bool jerry_enable_cpu_profiling (void);
 bool jerry_start_cpu_profiling (const char *path, double duration);
 bool jerry_stop_cpu_profiling (void);
 
+void jerry_heap_profiler_take_snapshot (const char *path);
```
Comment (yorkie) [inline deps/jerry/jerry-core/include/jerryscript-core.h]: `jerry_start_cpu_profiling` doesn't own a prefix like `cpu_profiler`, how about the shorter name `jerry_take_snapshot` here?

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h [hunk 15]
```diff
@@ -225,9 +226,7 @@ LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_FREEZE, "freeze")
 #if !defined (CONFIG_DISABLE_DATE_BUILTIN)
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_GET_DAY_UL, "getDay")
 #endif
-#if !defined (CONFIG_DISABLE_REGEXP_BUILTIN)
```
Comment (yorkie) [inline deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: This seems unrelated JerryScript problems?

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h [hunk 16]
```diff
@@ -501,6 +503,9 @@ LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_TO_LOCALE_STRING_UL, "toLocaleString")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_GET_MILLISECONDS_UL, "getMilliseconds")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SET_MILLISECONDS_UL, "setMilliseconds")
 #endif
+#if  !defined (CONFIG_DISABLE_ES2015_ARROW_FUNCTION) && defined (JERRY_HEAP_PROFILER)
```
Comment (yorkie) [inline deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: A unnecessary space before `!defined` :)

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h [hunk 17]
```diff
@@ -528,8 +533,14 @@ LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SET_UTC_MILLISECONDS_UL, "setUTCMilliseco
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_TO_LOCALE_DATE_STRING_UL, "toLocaleDateString")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_TO_LOCALE_TIME_STRING_UL, "toLocaleTimeString")
 #endif
+#if  defined (JERRY_HEAP_PROFILER)
```
Comment (yorkie) [inline deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: Ditto.

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h [hunk 18]
```diff
@@ -528,8 +533,14 @@ LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_SET_UTC_MILLISECONDS_UL, "setUTCMilliseco
...
+LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_EXTERNAL_FUNCTION, "<external function>")
+#endif
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_GET_OWN_PROPERTY_NAMES_UL, "getOwnPropertyNames")
 LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_PROPERTY_IS_ENUMERABLE_UL, "propertyIsEnumerable")
+#if defined (JERRY_HEAP_PROFILER)
+LIT_MAGIC_STRING_DEF (LIT_MAGIC_STRING_LEX_ENV, "<lexical environment>")
+#endif
```
Comment (yorkie) [inline deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: Why not grouping these `JERRY_HEAP_PROFILER` magic string together?

File: deps/jerry/jerry-core/lit/lit-magic-strings.ini [hunk 19]
```diff
@@ -263,3 +263,9 @@ LIT_MAGIC_STRING_GET_OWN_PROPERTY_NAMES_UL = "getOwnPropertyNames"
 LIT_MAGIC_STRING_PROPERTY_IS_ENUMERABLE_UL = "propertyIsEnumerable"
 LIT_MAGIC_STRING_GET_OWN_PROPERTY_DESCRIPTOR_UL = "getOwnPropertyDescriptor"
 LIT_MAGIC_STRING__FUNCTION_TO_STRING = "function(){/* ecmascript */}"
+LIT_MAGIC_STRING_EXTERNAL_FUNCTION = "<external function>"
+LIT_MAGIC_STRING_ARROW_FUNCTION = "<arrow function>"
```
Comment (yorkie) [inline deps/jerry/jerry-core/lit/lit-magic-strings.ini]: Keep consistent with LIT_MAGIC_STRING__FUNCTION_TO_STRING as the following:

```
LIT_MAGIC_STRING__EXTERNAL_FUNCTION_TO_STRING
LIT_MAGIC_STRING__ARROW_FUNCTION_TO_STRING
```

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 20]
```diff
@@ -0,0 +1,189 @@
+'use strict';
+
+/* jerry heap snapshot to V8 heap snapshot converter */
+
+var fs = require('fs');
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Should be `const`.

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+
+const STRING_BASE_OFFSET = 1; // first string is "<dummy string>"
+
+class SnapshotConverter {
+  constructor(jerrySnapshotPath, outputPath) {
+    var inputData = fs.readFileSync(jerrySnapshotPath, 'utf8');
+    this.jerrySnapshot = JSON.parse(inputData);
+    this.outputPath = outputPath;
+    this.strings = ["<dummy>"];
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Single quoted :p

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 22]
```diff
@@ -0,0 +1,189 @@
...
+    // any edge must appear after it's from node,
+    // so this.edgeMap should has edge.from 
+    this.edgeMap.get(edge.from).push(edge);
+  }
+    
+  parse() {
+    var elements = this.jerrySnapshot.elements;
+    var string_index = STRING_BASE_OFFSET;
+    for (var i=0; i<elements.length; i++) {
+      var element = elements[i];
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Should use `let` for block scoping :)

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 23]
```diff
@@ -0,0 +1,189 @@
...
+        'edge_count': this.edgesOutput.length/3,
+        'trace_function_count': 0
+      },
+      'nodes': this.nodesOutput,
+      'edges': this.edgesOutput,
+      'trace_function_infos': [],
+      'trace_tree': [],
+      'samples': [],
+      'strings': this.strings
+    };
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Could we put this big JSON to another JSON source, and require via `require('./snapshot-tmpl.json')`?

File: docs/api/Profiler.md [hunk 24]
```diff
@@ -34,6 +34,18 @@ The profiling data will be stored into a file ${process.cwd()}/Profile-${Date.no
...
+### profiler.takeSnapshot()
+* Returns {null}
+
+The `profiler.takeSnapshot()` methd take a Heap profiler file into ${process.cwd()}/Profile-${Date.now()}.
+
+### profiler.takeSnapshot(path)
+* `path` {String} profiling data file path
+* Returns {null}
+
+The `profiler.takeSnapshot()` methd take a Heap profiler file which path is `path`.
```
Comment (yorkie) [inline docs/api/Profiler.md]: JavaScript doesn't have override, we should merge these 2 ways into:

```
profiler.takeSnapshot([path])
```

File: docs/devs/Optimization-Tips.md [hunk 25]
```diff
@@ -55,10 +55,27 @@ Add Javascript code like docs/api/Profiler.md, Iot.js will generate a Profiler f
...
 ```
+
+## Heap Profiler
+Adding below arguments when building and running IoT.js will enable Heap Profiler.
+
+```shell
+$ ./tools/build.py --jerry-heap-profiler
+```
+
+Add Javascript code like docs/api/Profiler.md, Iot.js will generate a Jerry Heap Profiler file, say Profiler-123.
```
Comment (yorkie) [inline docs/devs/Optimization-Tips.md]: Javascript => JavaScript, and it's better that the docs/api/Profiler.md would be linkable :)

File: docs/devs/Optimization-Tips.md [hunk 26]
```diff
@@ -55,10 +55,27 @@ Add Javascript code like docs/api/Profiler.md, Iot.js will generate a Profiler f
...
+
+Users can generate v8 heap snapshot file.
+```shell
+node deps/jerry/tools/profiler/j2v8snap.js Profiler-123 v8.heapsnapshot
```
Comment (yorkie) [inline docs/devs/Optimization-Tips.md]: Add the prefix `$` for shell block :p

File: docs/devs/Optimization-Tips.md [hunk 27]
```diff
@@ -55,10 +55,27 @@ Add Javascript code like docs/api/Profiler.md, Iot.js will generate a Profiler f
...
+```
+
+Then open in Chrome browser following below site.
+https://developer.chrome.com/devtools/docs/heap-profiling
```
Comment (yorkie) [inline docs/devs/Optimization-Tips.md]: Make this to be clickable?

File: src/modules/iotjs_module_profiler.c [hunk 28]
```diff
@@ -27,6 +37,7 @@ JS_FUNCTION(StopProfiling) {
 
 jerry_value_t InitProfiler() {
   jerry_value_t profiler = jerry_create_object();
+  iotjs_jval_set_method(profiler, "takeSnapshot", TakeSnapshot);
   iotjs_jval_set_method(profiler, "startProfiling", StartProfiling);
   iotjs_jval_set_method(profiler, "stopProfiling", StopProfiling);
```
Comment (yorkie) [inline src/modules/iotjs_module_profiler.c]: Make these strings are magic ones?

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h (hunk 15 above)
Comment (algebrait) [reply deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: According to comment, lit-magic-strings.inc.h should be generated by gen-magic-strings.py from lit-magic-strings.ini, not by hand.
Seems upstream is not strictly following this rule.

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h (hunk 18 above)
Comment (algebrait) [reply deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: All code style problem in this file should be ignored.

File: deps/jerry/jerry-core/ecma/base/ecma-alloc.c (hunk 7 above)
Comment (algebrait) [reply deps/jerry/jerry-core/ecma/base/ecma-alloc.c]: Before this PR, sizeof (ecma_extended_object_t) - sizeof (ecma_object_t) == sizeof (uint64_t).
But I add a size field into ecma_extended_object_t, which make the assertion failed. This field will increase 4 bytes for every ext object. To save memory without profiler, I would wrap it by macro.

File: deps/jerry/jerry-core/lit/lit-magic-strings.inc.h (hunk 15 above)
Comment (algebrait) [reply deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: According to comment, lit-magic-strings.inc.h should be generated by gen-magic-strings.py from lit-magic-strings.ini, not by hand.
Seems upstream is not strictly following this rule.

Comment (yorkie) [reply deps/jerry/jerry-core/lit/lit-magic-strings.inc.h]: Okay, just created https://github.com/Rokid/ShadowNode/issues/283 for tracking this one, skip it now :p

File: deps/jerry/jerry-core/ecma/base/ecma-alloc.c (hunk 7 above)
Comment (yorkie) [reply deps/jerry/jerry-core/ecma/base/ecma-alloc.c]: > This field will increase 4 bytes for every ext object.

How about still doing assertion but like this:

```c
#ifndef JERRY_HEAP_PROFILER
JERRY_STATIC_ASSERT (sizeof (ecma_extended_object_t) - sizeof (ecma_object_t) <= sizeof (uint64_t),
                     size_of_ecma_extended_object_part_must_be_less_than_or_equal_to_8_bytes);
#else /* JERRY_HEAP_PROFILER */
JERRY_STATIC_ASSERT (sizeof (ecma_extended_object_t) - sizeof (ecma_object_t) - sizeof (EXT_OBJ_SIZE_TYPE) <= sizeof (uint64_t),
                     size_of_ecma_extended_object_part_must_be_less_than_or_equal_to_8_bytes);
#endif
```

File: deps/jerry/jerry-core/ecma/operations/ecma-objects.c (hunk 13 above)
Comment (yorkie) [reply deps/jerry/jerry-core/ecma/operations/ecma-objects.c]: Ping @algebrait

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 29]
```diff
@@ -0,0 +1,103 @@
+'use strict';
+
+/* jerry heap snapshot to V8 heap snapshot converter */
+
+let fs = require('fs');
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: This should be `const`.

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 30]
```diff
@@ -0,0 +1,103 @@
...
+
+const STRING_BASE_OFFSET = 1; // first string is "<dummy string>"
+
+class SnapshotConverter {
+  constructor(jerrySnapshotPath, outputPath) {
+    let inputData = fs.readFileSync(jerrySnapshotPath, 'utf8');
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Ditto.

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 31]
```diff
@@ -0,0 +1,103 @@
...
+      } else if (element.type === 'edge') {
+        this.addEdge(element);
+      }
+    }
+  }
+  
+  genNodeEdges(edges) {
+    for (let i = 0; i < edges.length; i++) {
+      let edge = edges[i];
+      this.edgesOutput.push (edge.edge_type);
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Remove the space after the function name :)

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 32]
```diff
@@ -0,0 +1,103 @@
...
+      let name;
+      name = this.stringMap.get(edge.name);
+      this.edgesOutput.push (name);
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Ditto.

File: deps/jerry/tools/profiler/j2v8snap.js [hunk 33]
```diff
@@ -0,0 +1,103 @@
...
+      // v8 node items is a flattern array,
+      // so to_node is to_node_id * node_size
+      this.edgesOutput.push (this.nodeMap.get(edge.to) * 6);
```
Comment (yorkie) [inline deps/jerry/tools/profiler/j2v8snap.js]: Ditto.
